File: LeetCode-Java/Copy List with Random Pointer.java
/**
 * Definition for singly-linked list with a random pointer.
 * class RandomListNode {
 *     int label;
 *     RandomListNode next, random;
 *     RandomListNode(int x) { this.label = x; }
 * };
 */
public class Solution {
    public RandomListNode copyRandomList(RandomListNode head) {
        HashMap<RandomListNode,RandomListNode> record = new HashMap<RandomListNode,RandomListNode>();
        RandomListNode start = head;
        RandomListNode newStart = null, pre = null;
        while(start != null)
        {
            RandomListNode tempNode = null;
            if(record.containsKey(start))
                tempNode = record.get(start);
            else
            {
                tempNode = new RandomListNode(start.label);
                record.put(start,tempNode);
            }
            if(pre != null)
                pre.next = tempNode;
            if(newStart == null)
                newStart = tempNode;
                
            RandomListNode randomNode = null;
            if(record.containsKey(start.random))
            {
                randomNode = record.get(start.random);
            }
            else if(start.random != null)
            {
                randomNode = new RandomListNode(start.random.label);
                record.put(start.random, randomNode);
            }
            tempNode.random = randomNode;
            pre = tempNode;
            start = start.next;
        }
        return newStart;
    }
}

File: Leetcode-cpp/copy_list_with_random_pointer.cpp
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        RandomListNode* current = head;
        while (current != nullptr) {
            RandomListNode* replica = new RandomListNode(current->label);
            replica->next = current->next;
            current->next = replica;
            current = current->next->next;
        }
        current = head;
        while (current != nullptr) {
            if (current->random != nullptr) current->next->random = current->random->next;
            current = current->next->next;
        }
        current = head;
        RandomListNode* result = new RandomListNode(-1);
        RandomListNode* replica_current = result;
        while (current != nullptr) {
            replica_current->next = current->next;
            replica_current = replica_current->next;
            current->next = current->next->next;
            current = current->next;
        }
        return result->next;
    }
};